File: external/HeimdallrForExtension/Heimdallr/Classes/Mach/Private/Backtrace/hmd_thread_backtrace.c
//
//  hmd_thread_backtrace.c
//  Heimdallr
//
//  Created by 谢俊逸 on 2019/2/19.
//

#include "hmd_thread_backtrace.h"
#include <libgen.h>
#include "pthread_extended.h"
#include <stdlib.h>
#include <string.h>
#include <sys/time.h>
#include <os/base.h>
#include <dispatch/dispatch.h>
#include <execinfo.h>

#include <mach/mach.h>
#include "HMDAsyncThread.h"
#include "HMDMacro.h"
#include "hmd_logger.h"
#include "hmd_mach.h"
#include "hmd_stack_cursor.h"
#include "hmd_stack_cursor_machine_context.h"
#include "hmd_stack_cursor_self_thread.h"
#include "HMDAsyncImageList.h"
#include "HMDCompactUnwind.hpp"
#include "HMDAsyncSymbolicator.h"
#include "hmd_symbolicator.h"
#include "HMDCPUUtilties.h"

typedef void(malloc_logger_t)(uint32_t type, uintptr_t arg1, uintptr_t arg2, uintptr_t arg3, uintptr_t result,
                              uint32_t num_hot_frames_to_skip);
extern malloc_logger_t *malloc_logger;
static malloc_logger_t *tmp_malloc_logger;

hmd_thread hmdbt_main_thread;

pthread_t hmdbt_main_pthread;

static unsigned long hmdbt_main_adress;

#pragma mark - Thread info

static bool hmdset_backtrace_thread_info(hmdbt_backtrace_t *bt) {
    if (!bt) {
        return false;
    }
    struct thread_extended_info thread_expend_info;
    kern_return_t kr = hmdthread_getExtendInfo(bt->thread_id, &thread_expend_info);
    if (kr == KERN_SUCCESS) {
        if (!(thread_expend_info.pth_flags & TH_FLAGS_IDLE))
            bt->thread_cpu_usage = ((float)thread_expend_info.pth_cpu_usage) / (float) TH_USAGE_SCALE * 100.0;
        bt->flags = thread_expend_info.pth_flags;
        bt->run_state = thread_expend_info.pth_run_state;
        bt->pth_curpri = thread_expend_info.pth_curpri;
        return true;
    }
    return false;
}

#pragma mark - Thread

bool hmdbt_all_threads(thread_act_array_t *thread_list, mach_msg_type_number_t *count) {
    if (task_threads(mach_task_self(), thread_list, count) != KERN_SUCCESS) {
        return false;
    }
    
    return true;
}

bool hmdbt_suspend_thread(thread_t thread) {
    // 关闭malloc监控回调
    // 可能会出现: 其他thread 在malloc_logger 回调中(其中有lock,并且进入lock), 随后调用hmd_suspend_environment,
    // 暂停了该线程，随后backtraceOfAllThreads 触发 malloc_logger, 导致死锁
    if (malloc_logger != NULL) {
        tmp_malloc_logger = malloc_logger;
        malloc_logger = NULL;
    }
    
    kern_return_t kr;
    bool rst = true;
    if ((kr = thread_suspend(thread)) != KERN_SUCCESS) {
        // Record the error and keep going.
        rst = false;
    }
    
    return rst;
}

bool hmdbt_suspend_threads(thread_act_array_t thread_list, mach_msg_type_number_t count, thread_t current_thread) {
    // 关闭malloc监控回调
    // 可能会出现: 其他thread 在malloc_logger 回调中(其中有lock,并且进入lock), 随后调用hmd_suspend_environment,
    // 暂停了该线程，随后backtraceOfAllThreads 触发 malloc_logger, 导致死锁
    if (malloc_logger != NULL) {
        tmp_malloc_logger = malloc_logger;
        malloc_logger = NULL;
    }
    
    kern_return_t kr;
    bool rst = true;
    for (mach_msg_type_number_t i = 0; i < count; i++) {
        thread_t thread = thread_list[i];
        if (thread != current_thread) {
            if ((kr = thread_suspend(thread)) != KERN_SUCCESS) {
                // Record the error and keep going.
                rst = false;
            }
        }
    }
    
    return rst;
}

bool hmdbt_resume_thread(thread_t thread) {
    if (tmp_malloc_logger != NULL) {
        malloc_logger = tmp_malloc_logger;
        tmp_malloc_logger = NULL;
    }
    
    kern_return_t kr;
    bool rst = true;
    if ((kr = thread_resume(thread)) != KERN_SUCCESS) {
        // Record the error and keep going.
        rst = false;
    }
    
    return rst;
}

bool hmdbt_resume_threads(thread_act_array_t thread_list, mach_msg_type_number_t count, thread_t current_thread) {
    kern_return_t kr;
    bool rst = true;
    for (mach_msg_type_number_t i = 0; i < count; i++) {
        thread_t thread = thread_list[i];
        if (thread != current_thread) {
            if ((kr = thread_resume(thread)) != KERN_SUCCESS) {
                // Record the error and keep going.
                rst = false;
            }
        }
    }
    
    if (tmp_malloc_logger != NULL) {
        malloc_logger = tmp_malloc_logger;
        tmp_malloc_logger = NULL;
    }
    
    return rst;
}

#pragma mark - Private

void HMD_NO_OPT_ATTRIBUTE _hmdbt_origin_backtrace_of_thread(hmd_thread thread, hmdbt_backtrace_t *backtrace, unsigned long skippedDepth, hmd_machine_context* machine_context, bool is_current_thread) {
    if (!backtrace->frames) {
        GCC_FORCE_NO_OPTIMIZATION return;
    }
    hmdthread_getName(thread, backtrace->name, HMDBT_MAX_NAME_LENGTH);
    hmd_stack_cursor stack_cursor;
    if (is_current_thread) {
        hmdsc_initSelfThread(&stack_cursor, (int)skippedDepth + 1);
    }
    else {
        hmdsc_initCursor(&stack_cursor, NULL, NULL);
        if (hmdmc_get_state_with_thread(thread, machine_context, false)) {
            hmdsc_initWithMachineContext(&stack_cursor, machine_context);
        }
    }
    
    stack_cursor.fast_unwind = true;
    
    size_t frame_count = 0;
    while (frame_count < HMDBT_MAX_FRAMES_COUNT && stack_cursor.advanceCursor(&stack_cursor)) {
        hmdbt_frame_t *frame = &(backtrace->frames[frame_count]);
        frame->address = stack_cursor.stackEntry.address;
        frame->stack_index = frame_count;
        frame_count++;
    }
    
    backtrace->frame_count = frame_count;
    backtrace->thread_id = (thread_t)thread;
    GCC_FORCE_NO_OPTIMIZATION return;
}

#pragma mark - Backtraces

unsigned long HMD_NO_OPT_ATTRIBUTE hmdbt_top_app_backtrace_addr_of_main_thread(unsigned long skippedDepth, bool suspend) {
    bool is_current_thread = (hmdthread_self() == hmdbt_main_thread);
    KSMC_NEW_CONTEXT(machine_context);
    size_t frame_count = 0;
    unsigned long addr = 0;
    hmd_async_image_list_set_reading(&shared_app_image_list, true);
    
    // suspend begin
    if (suspend && !is_current_thread) {
        hmdbt_suspend_thread((thread_t)hmdbt_main_thread);
    }
    
    hmd_stack_cursor stack_cursor;
    if (is_current_thread) {
        hmdsc_initSelfThread(&stack_cursor, (int)skippedDepth + 1);
    }
    else {
        hmdsc_initCursor(&stack_cursor, NULL, NULL);
        if (hmdmc_get_state_with_thread(hmdbt_main_thread, machine_context, false)) {
            hmdsc_initWithMachineContext(&stack_cursor, machine_context);
        }
    }
    
    stack_cursor.fast_unwind = true;
    
    while (frame_count < HMDBT_MAX_FRAMES_COUNT && stack_cursor.advanceCursor(&stack_cursor)) {
        uintptr_t current_addr = stack_cursor.stackEntry.address;
        if(current_addr > 0 && hmd_async_image_containing_address(&shared_app_image_list, current_addr)) {
            addr = stack_cursor.stackEntry.address;
            break;
        }
        frame_count++;
    }
    
    if (suspend && !is_current_thread) {
        hmdbt_resume_thread((thread_t)hmdbt_main_thread);
    }
    
    hmd_async_image_list_set_reading(&shared_app_image_list, false);
    GCC_FORCE_NO_OPTIMIZATION return addr;
}

void hmdbt_init_app_main_addr(void) {
    void(^mainAdressBlock)(void) = ^{
        void * main_backtrace[100];
        int len = backtrace(main_backtrace, 100);
        if (len > 2 && len <= 100){
            hmdbt_main_adress = (unsigned long)main_backtrace[len-2];
        }else{
#if DEBUG
            assert(0);
#endif
            hmdbt_main_adress = 0;
        }
    };
    if (pthread_main_np() == 0) {
        dispatch_async(dispatch_get_main_queue(), mainAdressBlock);
    }
    else {
        mainAdressBlock();
    }
}

unsigned long hmdbt_get_app_main_addr(void) {
#if DEBUG
    assert(hmdbt_main_adress != 0);
#endif
    GCC_FORCE_NO_OPTIMIZATION return hmdbt_main_adress;
}

hmdbt_backtrace_t * HMD_NO_OPT_ATTRIBUTE hmdbt_origin_backtrace_of_main_thread(unsigned long skippedDepth, bool suspend, bool fast) {
    hmdbt_backtrace_t *backtrace = calloc(1, sizeof(hmdbt_backtrace_t));
    if (fast) {
        backtrace->name = NULL;
    }
    else {
        backtrace->name = calloc(HMDBT_MAX_NAME_LENGTH, sizeof(char));
    }
    
    backtrace->frames = calloc(HMDBT_MAX_FRAMES_COUNT, sizeof(hmdbt_frame_t));
    backtrace->thread_id = (thread_t)hmdbt_main_thread;
    hmdset_backtrace_thread_info(backtrace);
    backtrace->thread_idx = 0;
    bool is_current_thread = (hmdthread_self() == hmdbt_main_thread);
    KSMC_NEW_CONTEXT(machine_context);
    
    // suspend begin
    if (suspend && !is_current_thread) {
        hmdbt_suspend_thread((thread_t)hmdbt_main_thread);
    }
    
    _hmdbt_origin_backtrace_of_thread(hmdbt_main_thread, backtrace, skippedDepth+1, machine_context, is_current_thread);
    
    if (suspend && !is_current_thread) {
        hmdbt_resume_thread((thread_t)hmdbt_main_thread);
    }
    
    // suspend over
    GCC_FORCE_NO_OPTIMIZATION return backtrace;
}

hmdbt_backtrace_t * HMD_NO_OPT_ATTRIBUTE hmdbt_origin_backtrace_of_thread(hmd_thread thread, unsigned long skippedDepth, bool suspend) {
    hmdbt_backtrace_t *backtrace = calloc(1, sizeof(hmdbt_backtrace_t));
    if (backtrace == NULL) {
        return NULL;
    }
    
    backtrace->name = calloc(HMDBT_MAX_NAME_LENGTH, sizeof(char));
    backtrace->frames = calloc(HMDBT_MAX_FRAMES_COUNT, sizeof(hmdbt_frame_t));
    backtrace->thread_id = (thread_t)thread;
    backtrace->thread_idx = 0;
    hmdset_backtrace_thread_info(backtrace);
    
    bool is_current_thread = (hmdthread_self() == thread);
    KSMC_NEW_CONTEXT(machine_context);
    // suspend begin
    if (suspend && !is_current_thread) {
        hmdbt_suspend_thread((thread_t)thread);
    }
    
    _hmdbt_origin_backtrace_of_thread(thread, backtrace, skippedDepth+1, machine_context, is_current_thread);
    
    if (suspend && !is_current_thread) {
        hmdbt_resume_thread((thread_t)thread);
    }
    
    // suspend over
    GCC_FORCE_NO_OPTIMIZATION return backtrace;
}

hmdbt_backtrace_t * HMD_NO_OPT_ATTRIBUTE hmdbt_origin_backtraces_of_all_threads(int *size, unsigned long skippedDepth, bool suspend, unsigned long max_thread_count) {
    if (size == NULL) {
        return NULL;
    }
    
    thread_act_array_t thread_list;
    mach_msg_type_number_t origin_thread_count = 0;
    hmdbt_all_threads(&thread_list, &origin_thread_count);
    mach_msg_type_number_t thread_count = origin_thread_count;
    if (max_thread_count > 0 && origin_thread_count > max_thread_count) {
        thread_count = (mach_msg_type_number_t)max_thread_count;
    }
    
    hmdbt_backtrace_t *backtraces = calloc(thread_count, sizeof(hmdbt_backtrace_t));
    if (backtraces == NULL) {
        return NULL;
    }
    
    thread_t current_thread = (thread_t)hmdthread_self();
    KSMC_NEW_CONTEXT(machine_context);
    
    for (int i = 0; i < thread_count; i++) {
        thread_t thread = thread_list[i];
        
        hmdbt_backtrace_t *backtrace = &(backtraces[i]);
        backtrace->name = calloc(HMDBT_MAX_NAME_LENGTH, sizeof(char));
        backtrace->frames = calloc(HMDBT_MAX_FRAMES_COUNT, sizeof(hmdbt_frame_t));
        backtrace->thread_id = thread;
        backtrace->thread_idx = i;
        hmdset_backtrace_thread_info(backtrace);
    }
    
    // suspend begin
    if (suspend) {
        hmdbt_suspend_threads(thread_list, thread_count, current_thread);
    }
    
    for (int i=0; i<thread_count; i++) {
        thread_t thread = thread_list[i];
        hmdbt_backtrace_t *backtrace = &(backtraces[i]);
        _hmdbt_origin_backtrace_of_thread(thread, backtrace, skippedDepth + 1, machine_context, current_thread == thread);
    }
    
    if (suspend) {
        hmdbt_resume_threads(thread_list, thread_count, current_thread);
    }
    
    for(size_t index = 0; index < origin_thread_count; index++) {
        mach_port_deallocate(mach_task_self(), thread_list[index]);
    }
    
    vm_deallocate(mach_task_self(), (vm_address_t)thread_list, sizeof(thread_act_t) * origin_thread_count);
    *size = thread_count;
    GCC_FORCE_NO_OPTIMIZATION return backtraces;
}

void hmdbt_dealloc_bactrace(hmdbt_backtrace_t **backtrace, int size) {
    if (backtrace == NULL || *backtrace == NULL) {
        return;
    }
    
    for (int i=0; i<size; i++) {
        hmdbt_backtrace_t *bt = &((*backtrace)[i]);
        if (bt->frames != NULL) {
            free(bt->frames);
            bt->frames = NULL;
        }
        
        if (bt->name != NULL) {
            free(bt->name);
            bt->name = NULL;
        }
    }
    
    free(*backtrace);
    *backtrace = NULL;
}

File: external/Heimdallr/Heimdallr/Classes/Mach/Common/Public/HMDThreadBacktrace.m
//
//  HMDThreadBacktrace.m
//  Heimdallr
//
//  Created by 刘诗彬 on 2018/1/25.
//

#import "HMDThreadBacktrace.h"
#import "HMDAsyncThread.h"
#import "hmd_mach.h"
#import "hmd_thread_backtrace.h"
#import "HMDMacro.h"
#import "hmd_symbolicator.h"
#import "HMDCompactUnwind.hpp"
#import "HMDThreadBacktraceFrame.h"
#import "hmd_crash_async_stack_trace.h"

extern hmd_thread hmdbt_main_thread;

typedef void(malloc_logger_t)(uint32_t type, uintptr_t arg1, uintptr_t arg2, uintptr_t arg3, uintptr_t result,
                              uint32_t num_hot_frames_to_skip);
extern malloc_logger_t *malloc_logger;

static HMDThreadBacktrace *construct_with_hmd_thread_backtrace(hmdbt_backtrace_t *bt) {
    HMDThreadBacktrace *backtrace = [[HMDThreadBacktrace alloc] init];
    NSMutableArray<HMDThreadBacktraceFrame *> *stackFrames = [[NSMutableArray alloc] init];
    if (bt->frames != NULL) {
        for (int i = 0; i < bt->frame_count; i++) {
            HMDThreadBacktraceFrame *frame = [[HMDThreadBacktraceFrame alloc] init];
            hmdbt_frame_t *originFrame = &(bt->frames[i]);
            frame.stackIndex = originFrame->stack_index;
            frame.address = originFrame->address;
            [stackFrames addObject:frame];
        }
    }
    
    hmd_async_stack_reading(true);
    hmd_async_stack_record_t *pre_bt = hmd_async_stack_trace_mach_thread(bt->thread_id);
    bool has_pre_stack = false;
    size_t async_times = 0;
    int pre_stack_index = (int)bt->frame_count;
    if (pre_bt != NULL) {
        if (pre_bt->valid) {
            has_pre_stack = true;
            async_times = pre_bt->async_times;
            for (size_t i=pre_bt->skip_length; i < pre_bt->length; i++){
                HMDThreadBacktraceFrame *frame = [[HMDThreadBacktraceFrame alloc] init];
                frame.address = (uintptr_t)(pre_bt->backtrace[i]);
                frame.stackIndex = pre_stack_index;
                [stackFrames addObject:frame];
                pre_stack_index++;
            }
        }
    }
    hmd_async_stack_reading(false);
    
    if(bt->name != NULL) {
        NSString *name = [NSString stringWithUTF8String:bt->name];
        if (has_pre_stack && strlen(pre_bt->thread_name) > 0) {
            name = [NSString stringWithFormat:@"%@ <= %@", name, [NSString stringWithUTF8String:pre_bt->thread_name]];
        }
        backtrace.name = name;
    }
    
    backtrace.threadID = bt->thread_id;
    backtrace.threadIndex = bt->thread_idx;
    backtrace.stackFrames = [stackFrames copy];
    backtrace.threadCpuUsage = bt->thread_cpu_usage;
    backtrace.async_times = (int)async_times;
    return backtrace;
}

int hmd_async_stack_trace_base_info_current_thread(hmd_async_stack_record_base_info_t * _Nonnull info) {
    hmd_async_stack_record_t *pre_bt = hmd_async_stack_trace_current_thread();
    if (pre_bt && pre_bt->valid && info) {
        info->pre_pthread = pre_bt->pre_pthread;
        info->pre_thread = pre_bt->pre_thread;
        info->thread = pre_bt->thread;
        info->pthread = pre_bt->pthread;
        
        strncpy(info->pre_thread_ids, pre_bt->pre_thread_ids, sizeof(info->pre_thread_ids));
        strncpy(info->thread_name, pre_bt->thread_name, sizeof(info->thread_name));
        memcpy(info->backtrace, pre_bt->backtrace, sizeof(info->backtrace));
        
        info->length = pre_bt->length;
        info->async_times = pre_bt->async_times;
        return 1;
    }
    return -1;
}

#pragma mark - HMDThreadBacktrace

@implementation HMDThreadBacktrace

+ (void)load {
    if (pthread_main_np() == 0) {
        dispatch_async(dispatch_get_main_queue(), ^{
            hmdbt_main_thread = hmdthread_self();
            hmdbt_main_pthread = pthread_self();
        });
    }
    else {
        hmdbt_main_thread = hmdthread_self();
        hmdbt_main_pthread = pthread_self();
    }
}

- (instancetype)init {
    self = [super init];
    if (self) {
        _threadID = 0;
        _threadIndex = 0;
        _threadCpuUsage = 0;
        _crashed = NO;
        _isSymbol = NO;
    }
    
    return self;
}

#pragma mark - Public

+ (thread_t)mainThread {
    return (thread_t)hmdbt_main_thread;
}

+ (thread_t)currentThread {
    return (thread_t)hmdthread_self();
}

+ (vm_address_t)getImageHeaderAddressWithName:(NSString *)name {
    __block vm_address_t header_adress = 0;
    hmd_enumerate_image_list_using_block(^(hmd_async_image_t *image, int index, bool *stop) {
        NSString *imageName = [[NSString stringWithUTF8String:image->macho_image.name] lastPathComponent];
        if ([name isEqualToString:imageName]) {
            header_adress = image->macho_image.header_addr;
            *stop = true;
        }
    });
    return header_adress;
}

+ (vm_address_t)getAppImageHeaderAddressWithName:(NSString *)name {
    __block vm_address_t header_adress = 0;
    hmd_enumerate_app_image_list_using_block(^(hmd_async_image_t *image, int index, bool *stop) {
        NSString *imageName = [[NSString stringWithUTF8String:image->macho_image.name] lastPathComponent];
        if ([name isEqualToString:imageName]) {
            header_adress = image->macho_image.header_addr;
            *stop = true;
        }
    });
    return header_adress;
}

+ (NSArray<HMDThreadBacktrace *> *)backtraceOfAllThreadsWithParameter:(HMDThreadBacktraceParameter *)parameter {
    // get origin backtrance
    int size = 0;
    hmdbt_backtrace_t *backtraces = hmdbt_origin_backtraces_of_all_threads(&size, parameter.skippedDepth+1, parameter.suspend, parameter.maxThreadCount);
    NSArray<HMDThreadBacktrace *>* backtraceList = [self createBacktraceList:backtraces
                                                                        size:size
                                                                   keyThread:parameter.keyThread
                                                                 symbolicate:parameter.needDebugSymbol];
    hmdbt_dealloc_bactrace(&backtraces, size);
    GCC_FORCE_NO_OPTIMIZATION return backtraceList;
}

+ (HMDThreadBacktrace * _Nullable)backtraceOfThreadWithParameter:(HMDThreadBacktraceParameter *)parameter {
    hmdbt_backtrace_t *bt;
    if (parameter.isGetMainThread) {
        bt = hmdbt_origin_backtrace_of_main_thread(parameter.skippedDepth + 1, parameter.suspend, false);
    }else{
        bt = hmdbt_origin_backtrace_of_thread(parameter.keyThread, parameter.skippedDepth + 1, parameter.suspend);
    }
    if (bt == NULL) {
        GCC_FORCE_NO_OPTIMIZATION return nil;
    }
    HMDThreadBacktrace *backtrace = construct_with_hmd_thread_backtrace(bt);
    backtrace.crashed = YES;
    if (parameter.needDebugSymbol) {
        [backtrace symbolicate:true];
    }
    
    hmdbt_dealloc_bactrace(&bt, 1);
    GCC_FORCE_NO_OPTIMIZATION return backtrace;
    
}

+ (NSArray<HMDThreadBacktrace *> *)backtraceOfAllThreadsWithKeyThread:(thread_t)keyThread
                                                          symbolicate:(BOOL)symbolicate
                                                         skippedDepth:(NSUInteger)skippedDepth
                                                              suspend:(BOOL)suspend
                                                       maxThreadCount:(NSUInteger)maxThreadCount {
    // get origin backtrance
    int size = 0;
    hmdbt_backtrace_t *backtraces = hmdbt_origin_backtraces_of_all_threads(&size, skippedDepth+1, suspend, maxThreadCount);
    NSArray<HMDThreadBacktrace *>* backtraceList = [self createBacktraceList:backtraces
                                                                        size:size
                                                                   keyThread:keyThread
                                                                 symbolicate:symbolicate];
    hmdbt_dealloc_bactrace(&backtraces, size);
    GCC_FORCE_NO_OPTIMIZATION return backtraceList;
}

+ (HMDThreadBacktrace * _Nullable)backtraceOfMainThreadWithSymbolicate:(BOOL)symbolicate
                                                skippedDepth:(NSUInteger)skippedDepth
                                                     suspend:(BOOL)suspend {
    hmdbt_backtrace_t *bt = hmdbt_origin_backtrace_of_main_thread(skippedDepth + 1, suspend, false);
    if (bt == NULL) {
        GCC_FORCE_NO_OPTIMIZATION return nil;
    }
    
    HMDThreadBacktrace *backtrace = construct_with_hmd_thread_backtrace(bt);
    if (symbolicate) {
        [backtrace symbolicate:true];
    }
    
    backtrace.crashed = YES;
    hmdbt_dealloc_bactrace(&bt, 1);
    GCC_FORCE_NO_OPTIMIZATION return backtrace;
}

+ (HMDThreadBacktrace * _Nullable)backtraceOfThread:(thread_t)thread
                              symbolicate:(BOOL)symbolicate
                             skippedDepth:(NSUInteger)skippedDepth
                                  suspend:(BOOL)suspend {
    hmdbt_backtrace_t *bt = hmdbt_origin_backtrace_of_thread(thread, skippedDepth + 1, suspend);
    if (bt == NULL) {
        GCC_FORCE_NO_OPTIMIZATION return nil;
    }
    
    HMDThreadBacktrace *backtrace = construct_with_hmd_thread_backtrace(bt);
    backtrace.crashed = YES;
    if (symbolicate) {
        [backtrace symbolicate:true];
    }
    
    hmdbt_dealloc_bactrace(&bt, 1);
    GCC_FORCE_NO_OPTIMIZATION return backtrace;
}

+ (NSArray <HMDThreadBacktrace *>*)createBacktraceList:(hmdbt_backtrace_t *)backtraces
                                                  size:(NSInteger)size
                                             keyThread:(thread_t)keyThread
                                           symbolicate:(BOOL)symbolicate {
    NSMutableArray<HMDThreadBacktrace *> *backtraceList = [[NSMutableArray alloc] initWithCapacity:size];
    for (int i = 0; i < size; i ++) {
        hmdbt_backtrace_t *bt = &(backtraces[i]);
        HMDThreadBacktrace *backtrace = construct_with_hmd_thread_backtrace(bt);
        if (symbolicate) {
            [backtrace symbolicate:true];
        }
        
        if (backtrace.threadID == keyThread) {
            backtrace.crashed = YES;
        }
        
        [backtraceList addObject:backtrace];
    }
    
    return [backtraceList copy];
}

+ (HMDThreadBacktraceFrame *)symbolicateForAddress:(uintptr_t)address{
    HMDThreadBacktraceFrame *frame = [[HMDThreadBacktraceFrame alloc] init];
    frame.address = address;
    [frame symbolicate:true];
    return frame;
}

- (void)symbolicate:(bool)needSymbolName {
    if (_isSymbol) {
        return;
    }
    
    _isSymbol = YES;
    hmd_setup_shared_image_list_if_need();
    for (HMDThreadBacktraceFrame *frame in self.stackFrames) {
        [frame symbolicate:needSymbolName];
    }
    
    return;
}

- (uintptr_t)topAppAddress {
    NSArray<HMDThreadBacktraceFrame*> *frames = _stackFrames;
    if (!(frames && frames.count > 0)) {
        return 0;
    }
    
    uintptr_t address = 0;
    hmd_setup_shared_image_list(); // 首次调用耗时
    hmd_async_image_list_set_reading(&shared_app_image_list, true);
    for (HMDThreadBacktraceFrame * _Nonnull frame in frames) {
        if (frame && frame.address > 0 && hmd_async_image_containing_address(&shared_app_image_list, frame.address)) {
            address = frame.address;
            break;
        }
    }
    
    hmd_async_image_list_set_reading(&shared_app_image_list, false);
    return address;
}

- (uintptr_t)bottomAppAddress {
    NSArray<HMDThreadBacktraceFrame*> *frames = _stackFrames;
    if (!(frames && frames.count > 0)) {
        return 0;
    }
    
    __block uintptr_t address = 0;
    hmd_setup_shared_image_list(); // 首次调用耗时
    hmd_async_image_list_set_reading(&shared_app_image_list, true);
    [frames enumerateObjectsWithOptions:NSEnumerationReverse usingBlock:^(HMDThreadBacktraceFrame * _Nonnull frame, NSUInteger idx, BOOL * _Nonnull stop) {
        if (frame && frame.address > 0 && hmd_async_image_containing_address(&shared_app_image_list, frame.address)) {
            address = frame.address;
            *stop = YES;
        }
    }];
    
    hmd_async_image_list_set_reading(&shared_app_image_list, false);
    return address;
}

#pragma mark HMDJSONable
- (NSDictionary *)jsonObject {
    NSMutableDictionary *data = [NSMutableDictionary
        dictionaryWithObjectsAndKeys:@(_threadIndex), @"thread_index", @(_threadID), @"thread_id", @(_threadCpuUsage), @"thread_cpu_usage", nil];

    if (_name) {
        [data setObject:_name forKey:@"name"];
    }

    NSArray *frames = [self.stackFrames valueForKey:@"jsonObject"];

    if (frames) {
        [data setObject:frames forKey:@"frames"];
    }
    return data;
}

- (NSString *)description {
    NSMutableString *string =
        [NSMutableString stringWithFormat:@"Thread %lu(%lu) ", (unsigned long)_threadIndex, _threadID];

    if (self.name) {
        [string appendFormat:@" name:%@", _name];
    }

    [string appendFormat:@" (cpu_usage: %.2f%%)", _threadCpuUsage];
    for (HMDThreadBacktraceFrame *frame in self.stackFrames) {
        [string appendFormat:@"\n%@", frame];
    }

    [string appendString:@"\n"];
    return string;
}

+ (instancetype _Nullable)backtraceWithPointerArray:(NSPointerArray * _Nonnull)pointerArray {
    NSUInteger count = pointerArray.count;
    if(count == 0) DEBUG_RETURN(nil);
    
    HMDThreadBacktrace *backtrace = HMDThreadBacktrace.alloc.init;
    NSMutableArray<HMDThreadBacktraceFrame *> *stackFrames = [NSMutableArray arrayWithCapacity:count];
    
    for(NSUInteger index = 0; index < count; index++) {
        HMDThreadBacktraceFrame *frame = HMDThreadBacktraceFrame.alloc.init;
        frame.stackIndex = index;
        frame.address = (uintptr_t)[pointerArray pointerAtIndex:index];
        [stackFrames addObject:frame];
    }
    backtrace.name = @"null";
    backtrace.threadID = 0;
    backtrace.threadIndex = 0;
    backtrace.stackFrames = stackFrames;
    backtrace.crashed = YES;
    
    return backtrace;
}

@end
